<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.52">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jeffrey Lewis &amp; Carson Stacy">
<meta name="dcterms.date" content="2025-01-13">

<title>Getting Started in R and RStudio</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="01_GettingStartedInR_files/libs/clipboard/clipboard.min.js"></script>
<script src="01_GettingStartedInR_files/libs/quarto-html/quarto.js"></script>
<script src="01_GettingStartedInR_files/libs/quarto-html/popper.min.js"></script>
<script src="01_GettingStartedInR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01_GettingStartedInR_files/libs/quarto-html/anchor.min.js"></script>
<link href="01_GettingStartedInR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01_GettingStartedInR_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01_GettingStartedInR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01_GettingStartedInR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01_GettingStartedInR_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="01_GettingStartedInR.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li><li><a href="01_GettingStartedInR.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Getting Started in R and RStudio</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jeffrey Lewis &amp; Carson Stacy </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 13, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="exercise-description" class="level2">
<h2 class="anchored" data-anchor-id="exercise-description">Exercise Description</h2>
<p>This activity is intended to familiarize you with using R and RStudio to analyze genomic data</p>
</section>
<section id="learning-outcomes" class="level2">
<h2 class="anchored" data-anchor-id="learning-outcomes">Learning Outcomes</h2>
<p>At the end of this exercise, you should be able to:</p>
<ul>
<li>Open, modify, and knit a code document file to a pdf/html output</li>
<li>Relate Quarto code documents to a traditional lab notebook</li>
<li>Run R commands in a Quarto file</li>
</ul>
</section>
<section id="using-r-and-rstudio" class="level2">
<h2 class="anchored" data-anchor-id="using-r-and-rstudio">Using R and RStudio</h2>
<p>This is a Quarto document. It uses a simple formatting syntax called markdown for authoring HTML, PDF, and MS Word documents. For more details on using Quarto, see <a href="https://quarto.org" class="uri">https://quarto.org</a>.</p>
<p>Think of this type of document as an electronic lab notebook for your data analysis.</p>
<p>When you click the <strong>knit</strong> button, a document will be generated that includes both content as well as the output of code chunks within the document. You can embed an R code chunk like this:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># preform basic calculations</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span><span class="sc">+</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print a statement usint the print() function</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"R code in a .qmd code chunk works just like a script"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "R code in a .qmd code chunk works just like a script"</code></pre>
</div>
</div>
</section>
<section id="r-packages" class="level2">
<h2 class="anchored" data-anchor-id="r-packages">R packages</h2>
<p>Packages are collections of functions that extend the capabilities of R. Packages have to be downloaded and installed before they can be used.</p>
<p>We can install and load packages using the following commands.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install a package</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"pak"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the package</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pak)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can load a package so we can use any functions it contains. Packages only need to be installed once, but they need to be loaded every time you start a new R session.</p>
<p>The <code>pak</code> package is a package management tool that makes it easier to install and load packages. The <code>pak</code> function from the <code>pak</code> package is a more modern way to install packages if they aren’t already installed and up to date.</p>
<p>We will use some packages below and later in this course, so let’s make sure you have them installed now.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># install readr and dplyr packages</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pak</span>(<span class="st">"readr"</span>, <span class="st">"dplyr"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="working-with-genomic-data" class="level2">
<h2 class="anchored" data-anchor-id="working-with-genomic-data">Working with Genomic Data</h2>
<p>R is a powerful tool for analyzing genomic data. Let’s download a genomic data file from GitHub for us to work with today.</p>
<p>Now, we can download the data file and load it into R.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co">#  access-data-01</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># assign url to a variable</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>DE_data_url <span class="ot">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/clstacy/GenomicAnalysis/main/Data/Msn24_EtOH/DE_genes_Msn24_EtOH.txt"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>One way we will load the file is using a GUI built into RStudio. - Copy the url above (without the quotation marks). - Click on the “Import Dataset” button in the Environment tab of RStudio. - Select “From Text (readr)” - Paste the url into the “File” field at the top of the window. - Click “update” for the file to be retrieved. - Choose “Tab” as the delimiter. - If things look good in the preview, click “Import” and the data will be loaded into RStudio.</p>
<p>Another way to load the data is using code, which is more reproducible. This function reads a tab-separated file into R.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load the readr package</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load in the file and save it as an object</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>DE_results_msn24_EtOH <span class="ot">=</span> <span class="fu">read_tsv</span>(DE_data_url)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 5756 Columns: 18
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr  (3): gene_id, common_name, annotation
dbl (15): logFC_YPS606_WT_etoh_response, pvalue_YPS606_WT_etoh_response, fdr...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>Note that this function uses the package <code>readr</code> (a part of the tidyverse package we loaded above). If you don’t have that package (1) installed and (2) loaded into your script, it won’t work.</p>
</section>
<section id="working-with-data-in-r" class="level2">
<h2 class="anchored" data-anchor-id="working-with-data-in-r">Working with data in R</h2>
<p>There are many rows and columns. We can use the <code>glimpse()</code> function from the <code>dplyr</code> package to get a better idea of what the data looks like.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load a very common processing package</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'dplyr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    filter, lag</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># glimpse() function shows how the data is structured</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(DE_results_msn24_EtOH)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 5,756
Columns: 18
$ gene_id                             &lt;chr&gt; "YMR105C", "YML100W", "YER053C", "…
$ common_name                         &lt;chr&gt; "PGM2", "TSL1", "PIC2", "NCE102", …
$ annotation                          &lt;chr&gt; "Phosphoglucomutase", "Large subun…
$ logFC_YPS606_WT_etoh_response       &lt;dbl&gt; 7.5999973, 7.7618280, 6.6940024, 0…
$ pvalue_YPS606_WT_etoh_response      &lt;dbl&gt; 9.40e-38, 1.04e-35, 3.03e-39, 2.17…
$ fdr_YPS606_WT_EtOH_response         &lt;dbl&gt; 3.26e-35, 1.54e-33, 2.07e-36, 4.57…
$ logFC_YPS606_msn24dd_EtOH_response  &lt;dbl&gt; 0.78481798, 0.60949852, 1.73591634…
$ pvalue_YPS606_msn24dd_EtOH_response &lt;dbl&gt; 3.430000e-06, 8.401730e-04, 4.6400…
$ fdr_YPS606_msn24dd_EtOH_response    &lt;dbl&gt; 7.420000e-06, 1.398507e-03, 2.8600…
$ logFC_WT.vs.msn24dd_EtOH_response   &lt;dbl&gt; -6.815179, -7.152329, -4.958086, -…
$ pvalue_WT.vs.msn24dd_EtOH_response  &lt;dbl&gt; 6.34e-32, 2.53e-30, 1.35e-27, 3.23…
$ fdr_WT.vs.msn24dd_EtOH_response     &lt;dbl&gt; 3.65e-28, 7.28e-27, 2.59e-24, 4.65…
$ logFC_WT.vs.msn24dd_unstressed      &lt;dbl&gt; -0.144061475, -0.365016862, -0.126…
$ pvalue_WT.vs.msn24dd_unstressed     &lt;dbl&gt; 0.350436027, 0.041423492, 0.425528…
$ fdr_WT.vs.msn24dd_unstressed        &lt;dbl&gt; 0.998531082, 0.998531082, 0.998531…
$ logFC_WT.vs.msn24dd_EtOH_absolute   &lt;dbl&gt; -6.959241, -7.517346, -5.084564, -…
$ pvalue_WT.vs.msn24dd_EtOH_absolute  &lt;dbl&gt; 8.55e-37, 2.04e-35, 3.06e-36, 7.93…
$ fdr_WT.vs.msn24dd_EtOH_absolute     &lt;dbl&gt; 1.64e-33, 1.96e-32, 3.52e-33, 2.85…</code></pre>
</div>
</div>
<p>We see in the output there are 5756 rows and 18 columns in the data. The same information should be available in the environment panel of RStudio</p>
</section>
<section id="looking-at-data-in-rstudio" class="level2">
<h2 class="anchored" data-anchor-id="looking-at-data-in-rstudio">Looking at Data in RStudio</h2>
<p>If we want to take a closer look at the data, we have a few options. To see just the first few lines we can run the following command:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(DE_results_msn24_EtOH)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
  gene_id common_name annotation   logFC_YPS606_WT_etoh…¹ pvalue_YPS606_WT_eto…²
  &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt;                         &lt;dbl&gt;                  &lt;dbl&gt;
1 YMR105C PGM2        Phosphogluc…                  7.60                9.40e-38
2 YML100W TSL1        Large subun…                  7.76                1.04e-35
3 YER053C PIC2        Mitochondri…                  6.69                3.03e-39
4 YPR149W NCE102      Protein inv…                  0.714               2.17e- 7
5 YKL035W UGP1        UDP-glucose…                  4.42                7.55e-34
6 YLR258W GSY2        Glycogen sy…                  7.52                2.20e-40
# ℹ abbreviated names: ¹​logFC_YPS606_WT_etoh_response,
#   ²​pvalue_YPS606_WT_etoh_response
# ℹ 13 more variables: fdr_YPS606_WT_EtOH_response &lt;dbl&gt;,
#   logFC_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   pvalue_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   fdr_YPS606_msn24dd_EtOH_response &lt;dbl&gt;,
#   logFC_WT.vs.msn24dd_EtOH_response &lt;dbl&gt;, …</code></pre>
</div>
</div>
<p>This can be difficult to look at. For looking at data similar to an Excel file, RStudio allows this by clicking on the name of the data.frame in the top right corner of the IDE. We can also view a file by typing <code>View(filename)</code>. To open the data in a new window, click the “pop out” button next to “filter” just above the opened dataset.</p>
</section>
<section id="exploring-the-data" class="level2">
<h2 class="anchored" data-anchor-id="exploring-the-data">Exploring the data</h2>
<p>This dataset includes the log fold changes of gene expression in an experiment testing the ethanol stress response for the YPS606 strain of <em>S. cerevisiae</em> and an <em>msn2/4ΔΔ</em> mutant. There are also additional columns of metadata about each gene. In later classes, we will cover the details included, but we can already start answering questions.</p>
<p><strong>Using RStudio, answer the following questions:</strong></p>
<ol type="1">
<li><p>How many genes are included in this study?</p></li>
<li><p>Which gene has the highest log fold change in the <em>msn2/4ΔΔ</em> mutant EtOH response?</p></li>
<li><p>How many HSP genes are differentially expressed (FDR &lt; 0.01) in unstressed conditions for the mutant?</p></li>
<li><p>Do the genes with the largest magnitude fold changes have the smallest p-values?</p></li>
<li><p>Which isoform of phosphoglucomutase is upregulated in response to ethanol stress, PGM1 or PGM2? Do you think <em>msn2/4</em> is responsible for this difference?</p></li>
</ol>
<p><strong>Render this file into a pdf or html file once you’re finished.</strong> If an error occurs, that will let you know there is any issue that needs to be resolved.</p>
<p>System information for reproducibility:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pander<span class="sc">::</span><span class="fu">pander</span>(<span class="fu">sessionInfo</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><strong>R version 4.3.3 (2024-02-29)</strong></p>
<p><strong>Platform:</strong> aarch64-apple-darwin20 (64-bit)</p>
<p><strong>locale:</strong> en_US.UTF-8||en_US.UTF-8||en_US.UTF-8||C||en_US.UTF-8||en_US.UTF-8</p>
<p><strong>attached base packages:</strong> <em>stats</em>, <em>graphics</em>, <em>grDevices</em>, <em>utils</em>, <em>datasets</em>, <em>methods</em> and <em>base</em></p>
<p><strong>other attached packages:</strong> <em>dplyr(v.1.1.4)</em>, <em>readr(v.2.1.5)</em> and <em>pak(v.0.8.0)</em></p>
<p><strong>loaded via a namespace (and not attached):</strong> <em>crayon(v.1.5.2)</em>, <em>vctrs(v.0.6.5)</em>, <em>cli(v.3.6.2)</em>, <em>knitr(v.1.45)</em>, <em>rlang(v.1.1.3)</em>, <em>xfun(v.0.42)</em>, <em>generics(v.0.1.3)</em>, <em>jsonlite(v.1.8.8)</em>, <em>bit(v.4.0.5)</em>, <em>glue(v.1.7.0)</em>, <em>htmltools(v.0.5.8.1)</em>, <em>hms(v.1.1.3)</em>, <em>fansi(v.1.0.6)</em>, <em>rmarkdown(v.2.26)</em>, <em>pander(v.0.6.5)</em>, <em>evaluate(v.0.23)</em>, <em>tibble(v.3.2.1)</em>, <em>tzdb(v.0.4.0)</em>, <em>fastmap(v.1.2.0)</em>, <em>yaml(v.2.3.8)</em>, <em>lifecycle(v.1.0.4)</em>, <em>compiler(v.4.3.3)</em>, <em>Rcpp(v.1.0.12)</em>, <em>htmlwidgets(v.1.6.4)</em>, <em>pkgconfig(v.2.0.3)</em>, <em>rstudioapi(v.0.16.0)</em>, <em>digest(v.0.6.34)</em>, <em>R6(v.2.5.1)</em>, <em>tidyselect(v.1.2.0)</em>, <em>utf8(v.1.2.4)</em>, <em>curl(v.6.0.1)</em>, <em>parallel(v.4.3.3)</em>, <em>vroom(v.1.6.4)</em>, <em>pillar(v.1.9.0)</em>, <em>magrittr(v.2.0.3)</em>, <em>bit64(v.4.0.5)</em> and <em>tools(v.4.3.3)</em></p>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>